### PR TITLE
Update linux BPF compiler to use pcap file for filter generation

### PR DESCRIPTION
**Pull request by bewing**
Running tcpdump to generate BPF bytecode requires the NET_ADMIN capability on the tcpdump binary to attach to the interface in promiscuous mode.  Instead, we can map the Linux ARPHD value to a pcap DLT value, and generate the appropriate PCAP file to pass through stdin. 

This approach can be useful when running scripts as a non-root, normal user.

### CODE REVIEW

**Comment by guedou** [comment]
Thanks for this PR. I really like the idea!

Could you add some relevant unit tests? For example, I would except that the previous and new ways to get the "raw" BPF filter return the same result.

I have some comments to make on the PR, but I think that it is better to have unit tests first.

**Comment by bewing** [comment]
I am happy to work on adding some unit tests.  I've not worked with UTScapy before, so it may take me some time to setup a valid test environment and get current tests running successfully.

**Comment by bewing** [comment]
I think the best solution to the new test failing in pypy would be to add an equality check method to the BPF C structs -- that way, we could assert(a == b), and it would work for the Cython Structures, or the pypy binary string.  Thoughts?

**Comment by bewing** [comment]
I am extremely puzzled by the mypy test failure -- could someone with some background knowledge on the CI system fill me in on what I'm missing?

**Comment by p-l-** [comment]
A quick question: I don't understand why you are adding so many constants to the Linux-specific code.

**Comment by bewing** [comment]
> A quick question: I don't understand why you are adding so many constants to the Linux-specific code.

The code is mostly trying to replicate the switch statement behavior from `map_arphrd_to_dlt` from https://github.com/the-tcpdump-group/libpcap/blob/master/pcap-linux.c#L3301, which is using the same constants from the Linux kernel and pcap headers.  I believe the dictionary that maps the values is easier to read and maintain if we are directly copying the named constants, instead of managing a dictionary that looks like:
```python
arphrd_to_dlt = {
    1: 1,
    23: 1,
    770: 107,
    ...
}
```
If there is a concern about the number of lines of code added, we could consider moving the constants to a separate file, or eliminating the constants altogether, but it would make updating/testing the lookup table more difficult.

**Comment by gpotter2** [comment]
Eventually, because tcpdump requires libpcap, the cleanest way would be to replace this whole shebang with a call  to `pcap_compile`.

File: scapy/arch/common.py [hunk 1]
```diff
@@ -26,6 +27,9 @@
 
 # BOOT
 
+FAKE_PCAP = ("d4 c3 b2 a1 02 00 04 00 00 00 00 00 00"
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
FAKE_PCAP = (b'\xd4\xc3\xb2\xa1\x02\x00\x04\x00\x00\x00\x00\x00\x00\x00\x00'
```

File: scapy/arch/common.py [hunk 2]
```diff
@@ -26,6 +27,9 @@
...
+             "00 00 00 ff ff 00 00 %02x 00 00 00").replace(' ', '')
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
             b'\x00\xff\xff\x00\x00%s\x00\x00\x00')
```

File: scapy/arch/common.py [hunk 3]
```diff
@@ -7,6 +7,7 @@
 Functions common to different architectures
 """
 
+import codecs
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
```

File: scapy/arch/common.py [hunk 4]
```diff
@@ -159,26 +163,39 @@ def get_bpf_pointer(tcpdump_lines):
...
     if not TCPDUMP:
         raise Scapy_Exception("tcpdump is not available. Cannot use filter !")
+    fake_pcap = None
+    tcpdump_opts = [
+        conf.prog.tcpdump,
+        "-ddd",
+        "-s", str(MTU),
+    ]
+    if iface_type:
+        fake_pcap = codecs.getdecoder('hex')(FAKE_PCAP % (iface_type, ))[0]
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
        fake_pcap = FAKE_PCAP % struct.pack('B', iface_type)
```

File: scapy/arch/common.py [hunk 5]
```diff
@@ -159,26 +163,39 @@ def get_bpf_pointer(tcpdump_lines):
...
+        tcpdump_opts.append("-r-")
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
        tcpdump_opts.extend(["-r", "-"])
```

File: scapy/arch/common.py [hunk 6]
```diff
@@ -159,26 +163,39 @@ def get_bpf_pointer(tcpdump_lines):
...
 
 
-def compile_filter(bpf_filter, iface=None):
+def compile_filter(bpf_filter, iface=None, iface_type=None):
     """Asks Tcpdump to parse the filter, then build the matching
     BPF bytecode using get_bpf_pointer.
     """
     if not TCPDUMP:
         raise Scapy_Exception("tcpdump is not available. Cannot use filter !")
+    fake_pcap = None
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
```

File: scapy/arch/common.py [hunk 7]
```diff
@@ -159,26 +163,39 @@ def get_bpf_pointer(tcpdump_lines):
...
+    tcpdump_opts = [
+        conf.prog.tcpdump,
+        "-ddd",
+        "-s", str(MTU),
+    ]
+    if iface_type:
+        fake_pcap = codecs.getdecoder('hex')(FAKE_PCAP % (iface_type, ))[0]
+        tcpdump_opts.append("-r-")
+    else:
+        tcpdump_opts.extend(
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
        fake_pcap = None
        tcpdump_opts.extend(
```

File: scapy/arch/common.py [hunk 8]
```diff
@@ -159,26 +163,39 @@ def get_bpf_pointer(tcpdump_lines):
...
+            tcpdump_opts,
+            stdin=subprocess.PIPE if fake_pcap else None,
             stdout=subprocess.PIPE,
             stderr=subprocess.PIPE
         )
     except OSError as ex:
         raise Scapy_Exception("Failed to attach filter: %s" % ex)
-    lines, err = process.communicate()
+    if fake_pcap:
+        lines, err = process.communicate(fake_pcap)
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
```

File: scapy/arch/common.py [hunk 9]
```diff
@@ -159,26 +163,39 @@ def get_bpf_pointer(tcpdump_lines):
...
+        process = subprocess.Popen(
+            tcpdump_opts,
+            stdin=subprocess.PIPE if fake_pcap else None,
             stdout=subprocess.PIPE,
             stderr=subprocess.PIPE
         )
     except OSError as ex:
         raise Scapy_Exception("Failed to attach filter: %s" % ex)
-    lines, err = process.communicate()
+    if fake_pcap:
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
    lines, err = process.communicate(input=fake_pcap)
```

File: scapy/arch/common.py [hunk 10]
```diff
@@ -159,26 +163,39 @@ def get_bpf_pointer(tcpdump_lines):
...
+        lines, err = process.communicate(fake_pcap)
+    else:
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
```

File: scapy/arch/common.py [hunk 11]
```diff
@@ -159,26 +163,39 @@ def get_bpf_pointer(tcpdump_lines):
...
+        lines, err = process.communicate()
```
Comment (p-l-) [inline scapy/arch/common.py]:
```suggestion
```

**Comment by gpotter2** [comment]
Done in https://github.com/secdev/scapy/issues/2336

**Pull request closed without merging**